src/devices/machine/cs8900a.cpp

```diff
@@ -0,0 +1,1522 @@
+#endif
+
+    return ret_val;
+}
+
+/* ------------------------------------------------------------------------- */
+/*    reading and writing IO register functions                             */
+
+/*
+These registers are currently fully or partially supported:
+
+CS8900_PP_ADDR_CC_RXCFG        0x0102 * # RW - 4.4.6.,  p. 52 - 0003 *
+CS8900_PP_ADDR_CC_RXCTL        0x0104 * # RW - 4.4.8.,  p. 54 - 0005 *
+CS8900_PP_ADDR_CC_LINECTL      0x0112 * # RW - 4.4.16., p. 62 - 0013 *
+CS8900_PP_ADDR_SE_RXEVENT      0x0124 * # R- - 4.4.7.,  p. 53 - 0004 *
+CS8900_PP_ADDR_SE_BUSST        0x0138 * # R- - 4.4.21., p. 67 - 0018 *
+CS8900_PP_ADDR_TXCMD           0x0144 * # -W - 4.5., p. 70 - 5.7., p. 98 *
+CS8900_PP_ADDR_TXLENGTH        0x0146 * # -W - 4.5., p. 70 - 5.7., p. 98 *
+CS8900_PP_ADDR_MAC_ADDR        0x0158 * # RW - 4.6., p. 71 - 5.3., p. 86 *
+                            0x015a
+                            0x015c
+*/
+
+/* ------------------------------------------------------------------------- */
+/* TX/RX buffer handling */
+
+void cs8900a_device::cs8900_write_tx_buffer(u8 value,int odd_address)
+{
+    /* write tx data only if valid buffer is ready */
+    if(tx_state != CS8900_TX_READ_BUSST) {
+#ifdef CS8900_DEBUG_WARN_RXTX
+        log_message(cs8900_log, "WARNING! Ignoring TX Write without correct Transmit Condition! (odd=%d,value=%02x)",
+                    odd_address,value);
+#endif
+        /* ensure correct tx state (needed if transmit < 4 was started) */
+        cs8900_set_tx_status(0,0);
+    } else {
+#ifdef CS8900_DEBUG_RXTX_STATE
+        if(tx_count==0) {
+            log_message(cs8900_log,"TX: write frame (length=%04x)",tx_length);
+        }
+#endif
+
+        /* always write LH, LH... to tx buffer */
+        u16 addr = tx_buffer;
+        if(odd_address) {
+            addr++;
+            tx_buffer += 2;
+        }
+        tx_count++;            
+        SET_PP_8(addr, value);
+        
+#ifdef CS8900_DEBUG_RXTX_DATA
+        log_message(cs8900_log, "TX: %04x/%04x: %02x (buffer=%04x,odd=%d)",
+                    tx_count,tx_length,value,addr,odd_address);
+#endif                
+
+        /* full frame transmitted? */
+        if(tx_count==tx_length) {
+#ifdef CS8900_DEBUG_FRAMES
+            log_message(cs8900_log, "cs8900_arch_transmit() called with:                 "
+                "length=%4u and buffer %s", tx_length,
+                debug_outbuffer(tx_length, &cs8900_packetpage[CS8900_PP_ADDR_TX_FRAMELOC])
+                );
+#endif
+
+            if(!tx_enabled) {
+#ifdef CS8900_DEBUG_WARN_RXTX
+                log_message(cs8900_log,"WARNING! Can't transmit frame (Transmitter is not enabled)!");
+#endif
+            } else {
+                /* send frame */
+#ifdef CS8900_DEBUG
+                printf("SENDING from buf %p len %d\n", &cs8900_packetpage[CS8900_PP_ADDR_TX_FRAMELOC], tx_length);
+#endif
+                send(&cs8900_packetpage[CS8900_PP_ADDR_TX_FRAMELOC],tx_length);
```

Comment:
Consider using the send_complete_cb() to update the transmitter state when sending frames, to somewhat simulate the delay of getting those bits onto the wire.

Reply:
As part of my latest changes, I did consider this.  I have not done so at this point because: (1) this is how the VICE/TFE code did it, and it's the only "successful" emulation of the CS8900A I have to go on (2) I haven't been able to determine from the data sheet on the CS8900A if the VICE/TFE code is actually wrong.  I'm happy to move the post-send side-effect code off into the `send_complete_cb()` callback and commit that if it "works for me", but I can't say that it's any more proper or not and I don't want to interfere too hard with a working implementation that I didn't have a close hand in writing myself.